src/routing_table/mod.rs

```diff
@@ -662,16 +654,15 @@ impl<T: Binary + Clone + Copy + Debug + Default + Hash + Xorable> RoutingTable<T
             return OwnMergeState::AlreadyMerged;
         }
         for prefix in merge_details.sections.keys() {
-            let compatible_with_ours = self.our_prefix.is_compatible(prefix);
-            // This may be a section which has been merged from multiple sections currently still
-            // in our RT, so fix up our RT first.
-            if merge_details.merge_prefix.is_compatible(prefix) &&
-               !self.sections.contains_key(prefix) && !compatible_with_ours {
-                self.merge(prefix);
-            }
-            // Add an empty section in the table.
-            if !compatible_with_ours {
-                let _ = self.sections.entry(*prefix).or_insert_with(HashSet::new);
+            if *prefix == self.our_prefix || self.sections.contains_key(prefix) {
+                continue; // Already in our routing table.
+            } else if self.our_prefix.is_compatible(prefix) ||
```

Comment:
We don't remove any prefixes here so I don't think there's a need to check for missing ones.
And the check that every prefix is a neighbour would fail, unfortunately, because of the way we do the merge in two steps: we _do_ insert prefixes here that aren't our neighbour yet, but will be after `finish_merging_own_section`.
We will need to clean up the merge process to fix that, but that's beyond the scope of this PR.